lib/rules/no-useless-assignment.js

```diff
@@ -0,0 +1,458 @@
+                 */
+                const subsequentSegmentData = {
+
+                    /**
+                     * Cache of subsequent segment information list that have already been iterated.
+                     * @type {SubsequentSegmentData[]}
+                     */
+                    results: [],
+
+                    /**
+                     * Subsequent segments that have already been iterated on. Used to avoid infinite loops.
+                     * @type {Set<CodePathSegment>}
+                     */
+                    subsequentSegments: new Set(),
+
+                    /**
+                     * Unexplored code path segment.
+                     * If additional iterations are needed, consume this information and iterate.
+                     * @type {CodePathSegment[]}
+                     */
+                    queueSegments: [...targetAssignment.segment.nextSegments]
+                };
+
+
+                /**
+                 * Gets the subsequent segments from the current segment.
+                 * @returns {Iterable<SubsequentSegmentData>} the subsequent segments
+                 */
+                function *getSubsequentSegments() {
+                    yield* subsequentSegmentData.results;
+
+                    while (subsequentSegmentData.queueSegments.length > 0) {
+                        const nextSegment = subsequentSegmentData.queueSegments.shift();
+
+                        if (subsequentSegmentData.subsequentSegments.has(nextSegment)) {
+                            continue;
+                        }
+                        subsequentSegmentData.subsequentSegments.add(nextSegment);
+
+                        const assignmentInSegment = allAssignments.find(otherAssignment => otherAssignment.segment === nextSegment);
+
+                        if (!assignmentInSegment) {
+
+                            /*
+                             * Stores the next segment to explore.
+                             * If `assignmentInSegment` exists,
+                             * we are guarding it because we don't need to explore the next segment.
+                             */
+                            subsequentSegmentData.queueSegments.push(...nextSegment.nextSegments);
+                        }
+
+                        /** @type {SubsequentSegmentData} */
+                        const result = {
+                            segment: nextSegment,
+                            assignment: assignmentInSegment
+                        };
+
+                        subsequentSegmentData.results.push(result);
+                        yield result;
+                    }
+                }
+
+
+                const readReferences = targetAssignment.variable.references.filter(reference => reference.isRead());
+
+                if (!readReferences.length) {
+
+                    // It is unused variable
+                    return;
```

Comment:
This is to avoid reporting on variables that are reported by `no-unused-vars`?

Reply:
I thought it shouldn't be reported in the new rule because it's not an unnecessary assignment, it's clearly an unnecessary variable. And it is reported by `no-unused-vars`.

Reply:
Can you add those details both as a comment here and in the documentation? I can see this being a bit confusing, but I agree with your rationale.